### PR TITLE
Update BootAuditTestApp to Correctly Test Load Option Variables

### PR DESCRIPTION
**Pull request by TaylorBeebe**
## Description

BootAuditTestApp tests the variable policy for the SysPrep####, PlatformRecovery####, and Driver#### variables by attempting to get and set the variables. When calling SetVariable with one of these three variables, VarCheckLib will sanity check the variable data and return EFI_INVALID_PARAMETER if the data is not a valid load option. Because we want to check if the variables are writable, we need to create properly formatted data before calling SetVariable.

This PR uses the DevicePath of the running EFI application to create a load option to properly test the writability of the boot option variables securely.

Fixes Issue https://github.com/microsoft/mu_plus/issues/252

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [x] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

Tested on Q35

## Integration Instructions

N/A

### CODE REVIEW

File: UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.c [hunk 1]
```diff
@@ -147,18 +260,19 @@ OsIndicationsSupportTest (
...
-  BASIC_TEST_CONTEXT  *Btc;
-  EFI_STATUS          Status;
-  UINTN               DataSize;
-  UINT32              Attributes;
-  UINT64              Data = 0x1122334455667788;
+  BASIC_TEST_CONTEXT            *Btc;
+  EFI_STATUS                    Status;
+  UINTN                         DataSize;
+  UINT32                        Attributes;
+  UINT64                        Data = 0x1122334455667788;
```
Comment (kuqin12) [inline UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.c:271]: Would we still run into the issue with invalid parameters if we try to set variable (i.e. sysprep) with this value?

File: UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.inf [hunk 2]
```diff
@@ -37,4 +37,12 @@
...
\ No newline at end of file
+  UnitTestLib
+  DevicePathLib
+  UefiBootManagerLib
+  UefiBootServicesTableLib
+  FileHandleLib
+
+[Protocols]
+  gEfiBlockIoProtocolGuid
+  gEfiSimpleFileSystemProtocolGuid
```
Comment (kuqin12) [inline UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.inf]: nit: new line for end of file.

**Review by kuqin12 — APPROVED** [review]

File: UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.c (hunk 1 above)
Comment (TaylorBeebe) [reply UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.c:271]: SysPrep####, PlatformRecovery####, and Driver#### have a different execution path in this test app (via checking the test context OptionType) and don't use this Data variable.

File: UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.inf (hunk 2 above)
Comment (TaylorBeebe) [reply UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.inf]: Added

File: UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.inf [hunk 3]
```diff
@@ -37,4 +37,12 @@
   UefiApplicationEntryPoint
   UefiLib
   UefiRuntimeServicesTableLib
-  UnitTestLib
\ No newline at end of file
+  UnitTestLib
+  DevicePathLib
```
Comment (makubacki) [inline UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.inf]: The library class items were sorted before. Could you sort these with the existing items?

File: UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.c [hunk 4]
```diff
@@ -31,28 +31,141 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
...
+#include <Library/DevicePathLib.h>
+#include <Library/UefiBootManagerLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Library/FileHandleLib.h>
 
-#define UNIT_TEST_APP_NAME     "BootAuditTestApp"
-#define UNIT_TEST_APP_VERSION  "1.0"
+#define UNIT_TEST_APP_NAME      "BootAuditTestApp"
+#define UNIT_TEST_APP_FILENAME  L"BootAuditTestApp.efi"
+#define UNIT_TEST_APP_VERSION   "1.0"
```
Comment (makubacki) [inline UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.c]: I suggest bumping the version. Easier to say "v1.1 has the change" later.

File: UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.c [hunk 5]
```diff
@@ -31,28 +31,141 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
...
+  // Locate all handles using the SFS protocol.
+  Status = gBS->LocateHandleBuffer (
+                  ByProtocol,
+                  &gEfiSimpleFileSystemProtocolGuid,
+                  NULL,
+                  &NumHandles,
+                  &HandleBuffer
+                  );
+
+  if (EFI_ERROR (Status) != FALSE) {
```
Comment (makubacki) [inline UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.c]: Is there a reason for the comparison to `FALSE` here but not in other checks in the function?

**Review by makubacki — APPROVED** [review]

File: UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.inf (hunk 3 above)
Comment (TaylorBeebe) [reply UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.inf]: Sorted

File: UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.c (hunk 4 above)
Comment (TaylorBeebe) [reply UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.c]: Bumped to v1.1

File: UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.c (hunk 5 above)
Comment (TaylorBeebe) [reply UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.c]: Copy-paste artifact -- removed

File: UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.c (hunk 1 above)
Comment (kuqin12) [reply UefiTestingPkg/AuditTests/BootAuditTest/UEFI/BootAuditTestApp.c:271]: Thanks for the explanation.

**Pull request merged**